### PR TITLE
Improve usage errors

### PR DESCRIPTION
**Pull request by hramezani**
Selected Reviewer: @samuelcolvin

### CODE REVIEW

File: docs/usage/errors.md [hunk 1]
```diff
@@ -3,99 +3,442 @@ with some suggestions on how to fix them.
 
 ## Decorator on missing field {#decorator-missing-field}
 
-TODO
+This error is raised when you define a decorator with an invalid field.
 
-## Dataclass not fully defined {#dataclass-not-fully-defined}
```
Comment (hramezani) [inline docs/usage/errors.md:8]: It's not in use after 3543649ee2f4318e525960c8554ebf5e8b98dfc7

File: docs/usage/errors.md [hunk 2]
```diff
@@ -119,14 +491,53 @@ The `field` argument is no longer available.
 
 ## V1 validator signature {#validator-v1-signature}
 
-TODO
+This error is raised when you use unsupported signature for V1 style validator.
```
Comment (tpdorsey) [inline docs/usage/errors.md:494]:
```suggestion
This error is raised when you use an unsupported signature for V1 style validator.
```

**Review by tpdorsey — COMMENTED** [review]
Minor suggested edits to docs.

**Review by samuelcolvin — COMMENTED** [review]
looks good, with @tpdorsey's suggestions and conflicts solved.

**Comment by samuelcolvin** [comment]
please update.

File: docs/usage/errors.md [hunk 3]
```diff
@@ -4,99 +4,442 @@ with some suggestions on how to fix them.
...
+    def check_a(cls, v: Any):
+        return v
+
+model = create_model('FooModel', a=(str, 'cake'), __base__=Model)
+```
 
 ## Discriminator no field {#discriminator-no-field}
 
-TODO
+This error is raised when a model in discriminated unions doesn't define discriminator field.
```
Comment (tpdorsey) [inline docs/usage/errors.md:40]:
```suggestion
This error is raised when a model in discriminated unions doesn't define a discriminator field.
```

File: docs/usage/errors.md [hunk 4]
```diff
@@ -4,99 +4,442 @@ with some suggestions on how to fix them.
...
+
+class Model(BaseModel):
+    pet: Union[Cat, Dog] = Field(..., discriminator='pet_type')
+    number: int
+```
 
 ## Discriminator alias type {#discriminator-alias-type}
 
-TODO
+This error is raised when you define a non-string alias on discriminator field.
```
Comment (tpdorsey) [inline docs/usage/errors.md:63]:
```suggestion
This error is raised when you define a non-string alias on a discriminator field.
```

File: docs/usage/errors.md [hunk 5]
```diff
@@ -120,14 +492,53 @@ The `field` argument is no longer available.
...
+
+    @validator('a')
+    def check_a(cls, value, foo):
+        return value
+```
+
+## Unrecognized field_validator signature {#validator-signature}
 
-## Unrecognized field_validator signature {#field-validator-signature}
+This error is raised when a `field_validator` or `model_validator` function has wrong signature.
```
Comment (tpdorsey) [inline docs/usage/errors.md:511]:
```suggestion
This error is raised when a `field_validator` or `model_validator` function has the wrong signature.
```

**Review by tpdorsey — APPROVED** [review]
Minor suggested edits.

**Pull request merged**